README.rst
=======
Dotconf
=======

Dotconf is an advanced configuration parser which allow nested sections to any
level, typed values in syntax, file include and so more. Dotconf is also
shipped with a powerful schema validation system.

The full documentation can be found here: http://dotconf.readthedocs.org

.. image:: https://www.ohloh.net/p/dotconf-python/widgets/project_thin_badge.gif
   :target: https://www.ohloh.net/p/dotconf-python?ref=sample


Features
--------

- Configuration format kept as simple as possible.
- Developer friendly APIs.
- Types are expressed in the syntax, so the parser can guess it without
  validation.
- Four primitive types: string, number, boolean, or list.
- More complex composite types can be created using the schema validation
  system.
- Nested section of any deep. Sections can take a special "argument" value
  (look at the example).
- Schema validation system, all schema can be defined using declarative or
  imperative way.
- External file include with globbing.
- Integration with argparse.
- Tests (only parser is covered yet)
- Released under MIT license.


Example
-------

This is an example of configuration file for an imaginary web server::


    daemon = True
    pidfile = '/var/run/myapp.pid'
    interface = '0.0.0.0:80'
    interface_ssl = '0.0.0.0:443'

    host 'example.org' {
        path '/' {
            rate_limit = 30
        }
    }

    host 'protected.example.org' {
        enable_ssl = yes

        path '/files' {
            enable_auth = yes
            user 'foo' {
                password = 'bar'
            }
        }
    }

You can access to each values using the developer friendly API::

    >>> from dotconf import Dotconf
    >>> parsed_conf = Dotconf.from_file('mywebserver.conf')
    >>> print parsed_conf.get('daemon', False)
    True


Even more exciting, you can create a validation schema to avoid you the
painful chore of manual configuration file validation::

    from dotconf.schema import many, once
    from dotconf.schema.containers import Section, Value
    from dotconf.schema.types import Boolean, Integer, Float, String

    # Schema definition:

    class UserSection(Section):
        password = Value(String())
        _meta = {'repeat': many, 'unique': True}

    class PathSection(Section):
        rate_limit = Value(Float(), default=0)
        enable_auth = Value(Boolean(), default=False)
        user = UserSection()

    class VirtualHostSection(Section):
        base_path = Value(String())
        enable_ssl = Value(Boolean(), default=False)
        path = PathSection()
        _meta = {'repeat': many, 'unique': True}

    class MyWebserverConfiguration(Section):
        daemon = Value(Boolean()default=False)
        pidfile = Value(String(), default=None)
        interface = Value(String(), default='127.0.0.1:80')
        interface_ssl = Value(String(), default='127.0.0.1:443')
        host = VirtualHostSection()

Then you can use the API exactly as if it was not validated::

    >>> from dotconf import Dotconf
    >>> from myconfschema import MyWebserverConfiguration
    >>> parsed_conf = Dotconf(conf, schema=MyWebserverConfiguration)
    >>> print 'daemon:', parsed_conf.get('daemon')
    daemon: True
    >>> for vhost in parsed_conf.subsections('host'):
    >>>     print vhost.args[0]
    >>>     if vhost.get('enable_ssl'):
    >>>         print '  SSL enabled'
    >>>     for path in vhost.subsections('path'):
    >>>         print '  ' + path.args[0]
    >>>         if path.get('enable_auth'):
    >>>             print '    Following users can access to this directory:'
    >>>             for user in path.subsections('user'):
    >>>                 print '     - ' + user.args[0]
    >>>
    example.org
      /
    protected.example.org
      SSL enabled
      /files
        Following users can access to this directory:
          - foo

Setup
-----

The fastest and more common way to install Dotconf is using pip::

    pip install dotconf

Debian
~~~~~~

If you use Debian, you can also use the Tecknet repositories. Add this lines
in your ``/etc/apt/source.list`` file::

    deb http://debian.tecknet.org/debian squeeze tecknet
    deb-src http://debian.tecknet.org/debian squeeze tecknet

Add the Tecknet repositories key in your keyring:

    # wget http://debian.tecknet.org/debian/public.key -O - | apt-key add -

Then, update and install::

    # aptitude update
    # aptitude install python-dotconf

Archlinux
~~~~~~~~~

If you use Archlinux, a Dotconf package is available in Aur::

    yaourt -S python2-dotconf


TODO
----

- More test.


Changelog
---------

v1.7 (not yet released)
~~~~~~~~~~~~~~~~~~~~~~~

-

v1.6 released on 09/12/12
~~~~~~~~~~~~~~~~~~~~~~~~~

This second stable release bring some bug fixes and features, the API has not
been broken. I also registered the project on travis-ci and I will try to
improve the test coverage for the next release.

Changes:

- Added Choice container
- Added a from_filename constructor the the Dotconf class
- Added encoding management (by default, files are parsed in UTF-8)
- Added continuous integration with travis
- Fixed bug with Float type validation
- Fixed an error when a section is included by an external file (thanks to
  DuanmuChun for its bug report and help to fix it)
- Fixed other minor bugs

New contributors:

- DuanmuChun (bug report and help to fix it)

v1.5 released on 14/04/2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~

First stable release of Dotconf has been released, development will now take
care of API compatibility. The project status has been changed to "Beta" on the
PYPI, and should be "Stable" on the next release if no major bug is found.
Packages will be updated for Debian and Archlinux, feel free to contact me if
you want to package it for your distro.

Changes:

- Added Eval, NamedRegex and RegexPattern types
- Added TypedArray container
- Fixed bug with scalar values from a singleton list in Value container
- Fixed argument validation in Section container
- Updated documentation (new tips and tricks section)

New contributors:

- Anaël Beutot (thanks for RegexPattern type and argument validation fix)

v0.4 released on 07/04/2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added debian package
- Added IPSocketAddress type
- Added Array container
- Added release procedure
- Fixed bug on IPAddress and IPNetwork types when ipaddr is missing
- Fixed documentation build

v0.3 released on 04/04/2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added IPAddress, IPNetwork, Regex and Url types
- Added min and max options on Integer type
- Added units on number parsing (42k == 42000)
- Fixed bug with validation of long numbers

v0.2 released on 03/04/2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Added argparse integration feature & documentation
- Cleanup

v0.1 released on 24/03/2012
~~~~~~~~~~~~~~~~~~~~~~~~~~~

- Initial version.


A note on versioning
--------------------

Dotconf use a two numbers X.Y versioning. The Y part is incremented by one on
each release. The X part is used as API compatibility indicator and will be
incremented each time the API is broken.


Contribute
----------

You can contribute to Dotconf through these ways:

- Main Git repository: https://idevelop.org/p/dotconf/source/tree/master/
- Bitbucket: https://bitbucket.org/NaPs/dotconf
- Github: https://github.com/NaPs/Dotconf

Feel free to contact me for any question/suggestion: <antoine@inaps.org>.


debian/changelog
dotconf (1.7~dev-1) stable; urgency=low

  * Development release

 -- Antoine Millet <antoine@inaps.org>  Mon, 10 Dec 2012 21:05:55 +0100

dotconf (1.6-1) stable; urgency=low

  * New upstream release: Dotconf 1.6

 -- Antoine Millet <antoine@inaps.org>  Sun, 09 Dec 2012 15:11:21 +0100

dotconf (1.5-1) stable; urgency=low

  * Initial package of Dotconf 1.5

 -- Antoine Millet <antoine@inaps.org>  Sat, 14 Apr 2012 22:51:09 +0200


dotconf/parser.py
""" Dotconf lexer and parser.
"""


from glob import glob

import ply.lex as lex
import ply.yacc as yacc

from dotconf.tree import ConfigSection, ConfigValue, Position


UNITS = {'k': 10 ** 3,
         'M': 10 ** 6,
         'G': 10 ** 9,
         'T': 10 ** 12,
         'P': 10 ** 15,
         'E': 10 ** 18,
         'Z': 10 ** 21,
         'Y': 10 ** 24,
         'Ki': 2 ** 10,
         'Mi': 2 ** 20,
         'Gi': 2 ** 30,
         'Ti': 2 ** 40,
         'Pi': 2 ** 50,
         'Ei': 2 ** 60,
         'Zi': 2 ** 70,
         'Yi': 2 ** 80}


class ParsingError(Exception):

    """ Error raised when a parsing error occurs.
    """

    def __init__(self, msg, position=None):
        super(ParsingError, self).__init__(msg)
        self.position = position


def default_external_opener(locator):
    """ The default locator used to open included external files.
    """
    parsed_externals = []
    for external in glob(locator):
        try:
            with open(external) as fexternal:
                external_data = fexternal.read()
        except IOError as err:
            raise ParsingError('Unable to open %s (%s)' % (external, err))
        parser = DotconfParser(external_data, debug=False, write_tables=False,
                               errorlog=yacc.NullLogger(), input_name=external,
                               external_opener=default_external_opener)
        parsed_externals.append(parser.parse())
    return parsed_externals


#
# Lexer
#

class DotconfLexer(object):

    """ Lexer for the DotConf format.

    :param \*\*kwargs: arguments to give to the ply lexer

    Usage example::

    >>> lexer = DotConfLexer()
    >>> lexer.input('test { key = yes }')
    >>> print lexer.next()
    """

    def __init__(self, encoding='utf-8', **kwargs):
        self._lexer = lex.lex(module=self, **kwargs)
        self._encoding = encoding

    #
    # Tokens definition
    #

    reserved = {'yes': 'YES', 'no': 'NO', 'include': 'INCLUDE'}
    reserved.update(dict(((kw, 'UNIT') for kw in UNITS)))
    tokens = ['LBRACE', 'RBRACE', 'NAME', 'TEXT', 'NUMBER',
              'ASSIGN', 'LIST_SEP'] + list(set(reserved.values()))

    t_LBRACE = '{'
    t_RBRACE = '}'
    t_ASSIGN = '='
    t_LIST_SEP = ','
    t_ignore = ' \t'

    def t_NAME(self, token):
        r'[a-zA-Z_][a-zA-Z0-9_-]*'
        token.type = self.reserved.get(token.value, 'NAME')
        # Handle the boolean case:
        if token.type == 'YES':
            token.value = True
        elif token.type == 'NO':
            token.value = False
        elif token.type == 'UNIT':
            token.value = UNITS[token.value]
        return token

    def t_TEXT(self, token):
        r'(["]([\\]["]|[^"]|)*["]|[\']([\\][\']|[^\'])*[\'])'
        value = token.value[1:-1].replace('\\' + token.value[0], token.value[0])
        token.value = value.decode(self._encoding)
        # Count the lines in the string:
        token.lexer.lineno += value.count('\n')
        return token

    def t_NUMBER(self, token):
        r'[-+]?[0-9]+(\.[0-9]+)?'
        if token.value.isdigit():
            token.value = int(token.value)
        else:
            token.value = float(token.value)
        return token

    def t_EOL(self, token):
        r'[\n]+'
        token.lexer.lineno += len(token.value)

    def t_COMMENT(self, token):
        r'[#].*'

    def t_error(self, token):
        position = Position(self._input_name,
                            self._lexer.lineno,
                            self._lexer.lexpos)
        raise ParsingError('Illegal character %r' % token.value[0], position)


    #
    # Public methods
    #

    def column(self, lexpos):
        """ Find the column according to the lexpos.
        """
        # This code is taken from the python-ply documentation
        # see: http://www.dabeaz.com/ply/ply.html section 4.6
        last_cr = self._current_input.rfind('\n', 0, lexpos)
        if last_cr < 0:
            last_cr = 0
        column = (lexpos - last_cr)
        return column


    #
    # Bindings to the internal _lexer object
    #

    def input(self, input):
        self._current_input = input
        return self._lexer.input(input)

    def __getattr__(self, name):
        attr = getattr(self._lexer, name)
        if attr is None:
            raise AttributeError("'%s' object has no attribute '%s'" % (self, name))
        else:
            return attr

#
# Parser
#

class DotconfParser(object):

    """ Parser for the Dotconf format.
    """

    tokens = DotconfLexer.tokens

    def __init__(self, input, **kwargs):
        self._input = input
        self._input_name = kwargs.pop('input_name', '<unknown>')
        self._external_opener = kwargs.pop('external_opener',
                                           default_external_opener)
        self._lexer = kwargs.pop('lexer', DotconfLexer())
        self._parser = yacc.yacc(module=self, **kwargs)
        self._old_line = 0

    def _check_line(self, current, lineno, pos, token):
        if self._old_line == current:
            pos = Position(self._input_name, lineno, pos)
            raise ParsingError('Syntax error near of "%s", '
                               'newline missing?' % token, pos)
        else:
            self._old_line = current

    #
    # Rules
    #

    start = 'top'

    def p_top(self, p):
        """top : section_content"""
        section = ConfigSection('__top__')
        for child in p[1]:
            if isinstance(child, ConfigSection):
                child.parent = section
            section.register(child)
        p[0] = section

    def p_assignation(self, p):
        """assignment : NAME ASSIGN value
                      | NAME ASSIGN list"""
        position = Position(self._input_name,
                            p.lineno(3),
                            p.lexer.column(p.lexpos(3)))
        value = ConfigValue(p[1], p[3], position=position)
        p[0] = value

    def p_value(self, p):
        """value : TEXT
                 | YES
                 | NO
                 | number"""
        p[0] = p[1]

    def p_number(self, p):
        """number : NUMBER
                  | NUMBER UNIT"""
        if len(p) == 2:
            p[0] = p[1]
        else:
            p[0] = p[1] * p[2]

    #
    # List management rules:
    #

    def p_list(self, p):
        """list : value LIST_SEP list_next
                | value LIST_SEP"""
        if len(p) == 3:
            p[0] = [p[1]]
        elif len(p) == 4:
            p[3].insert(0, p[1])
            p[0] = p[3]

    def p_list_next(self, p):
        """list_next : value
                     | value LIST_SEP
                     | value LIST_SEP list_next"""
        if len(p) in (2, 3):
            p[0] = [p[1]]
        elif len(p) == 4:
            p[3].insert(0, p[1])
            p[0] = p[3]

    #
    # Sections:
    #

    def p_section_content_empty(self, p):
        """section_content : empty"""
        p[0] = []

    def p_section_content_assignation(self, p):
        """section_content : section_content assignment
                           | section_content section"""
        self._check_line(p.lexer.lineno, p.lineno(2),
                         self._lexer.column(p.lexpos(2)), p[2].name)
        p[1].append(p[2])
        p[0] = p[1]

    def p_section_content_include(self, p):
        """section_content : section_content INCLUDE TEXT"""
        for external in self._external_opener(p[3]):
            p[1] += list(external.iterflatchildren())
        p[0] = p[1]

    def p_section(self, p):
        """section : NAME LBRACE section_content RBRACE
                   | NAME section_args LBRACE section_content RBRACE"""
        name = p[1]
        if len(p) == 5:
            args = None
            section_content = p[3]
        else:
            args = p[2]
            section_content = p[4]
        column = p.lexer.column(p.lexpos(1))
        position = Position(self._input_name, p.lineno(1), column)
        section = ConfigSection(name, args=args, position=position)
        for child in section_content:
            if isinstance(child, ConfigSection):
                child.parent = section
            section.register(child)
        p[0] = section

    def p_section_args(self, p):
        """section_args : value LIST_SEP section_args
                        | value"""
        if len(p) == 2:
            position = Position(self._input_name,
                                p.lineno(1),
                                p.lexer.column(p.lexpos(1)))
            p[0] = ConfigValue('<args>', [p[1]], position=position)
        else:
            p[3].value.insert(0, p[1])
            p[0] = p[3]

    def p_empty(self, p):
        """empty :"""
        pass

    def p_error(self, token):
        column = self._lexer.column(token.lexpos)
        pos = Position(self._input_name, token.lineno, column)
        raise ParsingError('Syntax error near of "%s"' % token.value, pos)


    #
    # Bindings to the internel _parser object
    #

    def parse(self):
        return self._parser.parse(self._input, self._lexer, tracking=True)

    def __getattr__(self, name):
        attr = getattr(self._parser, name)
        if attr is None:
            raise AttributeError("'%s' object has no attribute '%s'" % (self, name))
        else:
            return attr
